src/data/nat/choose.lean

```diff
@@ -100,8 +101,142 @@ theorem add_pow [comm_semiring α] (x y : α) (n : ℕ) :
 (commute.all x y).add_pow n
 
 /-- The sum of entries in a row of Pascal's triangle -/
-theorem sum_range_choose (n : ℕ) : 
+theorem sum_range_choose (n : ℕ) :
   ∑ m in range (n + 1), choose n m = 2 ^ n :=
 by simpa using (add_pow 1 1 n).symm
 
+/-
+Facts about more specific binomial coefficients and their sums
+-/
+
+private lemma succ_not_le {m : nat} (t : m.succ ≤ m) : false :=
+begin
+  induction m, linarith,
+  apply m_ih,
+  exact nat.lt_succ_iff.mp t,
+end
+
+lemma sum_range_split_halfway (f : nat → nat) (m : nat) :
+  ∑ i in range (2 * m + 2), f i = (∑ i in range (m + 1), f i) + ∑ i in (Ico (m + 1) (2 * m + 2)), f i :=
+begin
+  induction m with m,
+  {simp, refl},
+  {
+    rw sum_range_succ f (2 * (m + 1) + 1),
+    rw sum_range_succ f (2 * (m + 1)),
+    have s : 2 * m + 2 = 2 * (m + 1), refl,
+    rw s at m_ih,
+    rw m_ih,
+    clear m_ih,
+    rw sum_range_succ f (m + 1),
+    simp,
+    rw add_comm (∑ i in range (m + 1), f i) (∑ i in Ico (m + 2) (2 * (m + 1) + 2), f i),
+    rw <- add_assoc (f (m + 1)) _ _,
+    rw (@sum_eq_sum_Ico_succ_bot _ _ (m + 1) (2 * (m + 1) + 2) ( by linarith ) f).symm,
+    rw @sum_Ico_succ_top _ _ m.succ (2 * m.succ + 1) ( by linarith ) f,
+    rw @sum_Ico_succ_top _ _ m.succ (2 * m.succ) ( by linarith ) f,
+    ring,
+  }
+end
+
+-- This lemma exists only so that we can instantiate it with `i = m + 1`.
+private lemma reflect_sum_lemma (i : nat) (m : nat) (i_bound : i ≤ m + 1) (f : nat → nat) (reflects : ∀ x ≤ 2 * m + 1, f x = f (2 * m + 1 - x)) :
+  ∑ j in (Ico (m + 1 - i) (m + 1)), f j = ∑ j in Ico (m + 1) (m + 1 + i), f j :=
```

Comment:
I think `finset.sum_bij` is for this kind of thing.

Reply:
I've rewritten this mostly to use `finset.sum_bij`, but actually I think it's coming out much worse; it's now four really grotty proofs of the easy-and-obvious hypotheses of `finset.sum_bij`, rather than a single proof with an obvious statement. When the bijection is implicit, I think you get a nicer induction (because this particular bijection is itself very nice).

Reply:
Easy and obvious side conditions sound good! They should have easy and obvious proofs. The main difficulty is working in `nat` rather than `int`. I'm not up to speed on the best way to deal with these goals: @kbuzzard ? @ChrisHughes24 ?

Reply:
I'd like to have this lemma as a non-private theorem in `big_operators` in a more general case. I mean something like `∑ j in Ico a b, f j = ∑ j in Ico (c - b) (c - a), f (c - j - 1)`.

Reply:
I do think there should be a different version of `sum_bij` asking for an inverse function instead of bijectivity. Then the proof obligations will be equalities which should be easier to kill with tactics than injectivity or surjectivity proofs.

Reply:
Shall we get all the variations of `Ico`, like `Icc` as well. It's a ton more lemmas, but the proofs are nicer.

Reply:
Note that we have `prod_image`.

Reply:
I've rewritten this as a proof of approximately the same length but using `sum_bij` as expressed in #2709 .

Reply:
I PR'd a generic "reverse an `Ico`" lemma, see #2740.